src/clj/puppetlabs/services/jruby/jruby_internal.clj

```diff
@@ -139,39 +146,42 @@
 (schema/defn ^:always-validate
   cleanup-pool-instance!
   "Cleans up and cleanly terminates a JRubyInstance and removes it from the pool."
-  [{:keys [scripting-container pool] :as instance} :- JRubyInstance]
+  [{:keys [scripting-container pool] :as instance} :- jruby-schemas/JRubyInstanceSchema
+   shutdown-fn :- IFn]
   (.unregister pool instance)
-  ;; TODO: need to add support for a callback hook, so that consumers like
-  ;; puppet-server can do their own cleanup.
-  ;(.terminate jruby-puppet)
+  (shutdown-fn instance)
   (.terminate scripting-container)
   (log/infof "Cleaned up old JRubyInstance with id %s." (:id instance)))
 
 (schema/defn ^:always-validate
-  create-pool-instance! :- JRubyInstance
+  create-pool-instance! :- jruby-schemas/JRubyInstanceSchema
   "Creates a new JRubyInstance and adds it to the pool."
   [pool :- jruby-schemas/pool-queue-type
    id :- schema/Int
    config :- jruby-schemas/JRubyConfig
    flush-instance-fn :- IFn]
-  (let [{:keys [ruby-load-path gem-home compile-mode]} config]
+  (let [{:keys [ruby-load-path gem-home compile-mode lifecycle]} config
+        init-fn (:initialize lifecycle)
+        init-env-vars-fn (:initialize-env-variables lifecycle)]
```

Comment:
Yep, he did, the problem is it doesn't work :(. It seems like as soon as you call `.runScriptlet` that it does something special to set the environment variables, and then while you can call `.setEnvironment` after that, it doesn't get plumbed all the way through. :/